Initial documentation on Multiple Control Planes (#7035)
istio/istio.io: content/en/docs/setup/upgrade/index.md

---
title: Upgrade Istio
description: Upgrade or downgrade Istio.
weight: 25
keywords: [kubernetes,upgrading]
---

## Canary upgrades

> The recommended upgrade method is to do a canary deployment of the control plane,

Reviewer: I think this lead-in could be a little more clear. How about:
```
Upgrading Istio can be done by first running a canary deployment of the new control plane, allowing you
to monitor the effect of the upgrade with a small percentage of the workloads, before migrating all of the
traffic to the new version. This is much safer than doing an in place upgrade and is the recommended upgrade method.

The recommended upgrade method is to do a canary deployment of the control plane,
and slowly migrate traffic to the new version. This allows making safe changes
and monitor the impact on the mesh before moving all traffic over.
```

and slowly migrate traffic to the new version. This allows making safe changes
and monitor the impact on the mesh before moving all traffic over.

> To support canary deployments, Istio has an installation setting named `revision`.

Reviewer: ```suggestion
When installing istio, the `revision` installation setting can be used to deploy multiple independent control planes at the same time. A canary version of an upgrade can be started by installing the new Istio version's control plane next to the old one, using a different `revision` setting.
```

> Using this option, you can deploy independent control planes.

Reviewer: ```suggestion
```

> Each revision will have its own Deployment, Service, etc.

Reviewer: ```suggestion
Each revision is a full Istio control plane implementation with its own `Deployment`, `Service`, etc.
```

> Each revision will have its own Deployment, Service, etc.

Author: Its everything to configure a control plane. a deployment, service, configmaps, HorizontalPodAutoscalers, .... Basically a full duplicate of the install

### Control plane

To install a new revision, named `canary` for example, set the `revision` field:
`istioctl install --set revision=canary`.

> After this runs, see there are two deployments running side-by-side:

Reviewer: ```suggestion
After running the command, you will have two control plane deployments running side-by-side:
```

> Once this has run, we can see we have two deployments running side-by-side:

Reviewer 2: ```suggestion
After this runs, see there are two deployments running side-by-side:
```

{{< text bash >}}
$ kubectl get pods -n istio-system
NAME                                    READY   STATUS    RESTARTS   AGE
istiod-786779888b-p9s5n                 1/1     Running   0          114m
istiod-canary-6956db645c-vwhsk          1/1     Running   0          1m
{{< /text >}}

### Data plane

> Just installing the new revision will have no impact on our existing proxies. In order to upgrade these,

Reviewer 2: ```suggestion
Simply installing the new revision has no impact on the existing proxies. To upgrade these,
```

> configure them to point to the new control plane. This is controlled during sidecar injection

Reviewer: ```suggestion
you must configure them to point to the new control plane. This is controlled during sidecar injection
```

> we need to configure them to point to the new control plane. This is controlled during sidecar injection

Reviewer 2: ```suggestion
configure them to point to the new control plane. This is controlled during sidecar injection
```

> based on the namespace label `istio.io/rev`. To upgrade the namespace `test-ns`, relabel it

Reviewer: ```suggestion
based on the namespace label `istio.io/rev`.

To upgrade proxies in namespace `test-ns` to use the new control plane, set the label to the canary revision:
```

> based on the namespace label `istio.io/rev`. If we want to upgrade our namespace `test-ns`, we can relabel it

Reviewer 2: ```suggestion
based on the namespace label `istio.io/rev`. To upgrade the namespace `test-ns`, relabel it
```

> to point to the canary revision: `kubectl label namespace default istio-injection- istio.io/rev=canary`. Note the removal

Reviewer: ```suggestion

{{< text bash >}}
$ kubectl label namespace default istio-injection- istio.io/rev=canary
{{< /text >}}

Note that the above command is also removing the `istio-injection` label because otherwise
it would take precedence over the `istio.io/rev` label (for backward compatibility).
```

> of the `istio-injection` label, which takes precedence over the `istio.io/rev` label

Reviewer: ```suggestion
```

> for backwards compatibility.

Reviewer: ```suggestion
```

> After the namespace updates, restart the pods to trigger a re-injection. One way to do

Reviewer: ```suggestion
After the namespace updates, you need to restart the pods to trigger re-injection. One way to do
```

> to point to the canary revision: `kubectl label namespace default istio-injection- istio.io/rev=canary`. Note that
> we are also removing the `istio-injection` label, which will take precedence over the `istio.io/rev` label
> for backwards compatibility.

Reviewer 2: ```suggestion
to point to the canary revision: `kubectl label namespace default istio-injection- istio.io/rev=canary`. Note the removal
of the `istio-injection` label, which takes precedence over the `istio.io/rev` label
for backwards compatibility.
```

> this is using `kubectl rollout restart deployment -n test-ns`.

Reviewer: ```suggestion
this is using:

{{< text bash >}}
$ kubectl rollout restart deployment -n test-ns
{{< /text >}}
```

> Once our namespace has been updated, we need to restart our pods to trigger a re-injection. One way to do

Reviewer 2: ```suggestion
After the namespace updates, restart the pods to trigger a re-injection. One way to do
```

> When the pod is re-injected, it will be configured to point to the `istiod-canary` control plane. You can verify this by

Reviewer: ```suggestion
When the pods are re-injected, they will be configured to point to the `istiod-canary` control plane. You can verify this by
```

> looking at the pod labels. For example, `kubectl get pods -n test-ns -l istio.io/rev=canary` will show all

Reviewer: ```suggestion
looking at the pod labels.
For example, the following command will show all the pods using the `canary` revision:

{{< text bash >}}
$ kubectl get pods -n test-ns -l istio.io/rev=canary
{{< /text >}}
```

> When the pod is re-injected, it will be configured to point to the `istiod-canary` control plane. We can verify this by

Reviewer 2: ```suggestion
When the pod is re-injected, it will be configured to point to the `istiod-canary` control plane. You can verify this by
```

> pods using the `canary` revision.

Reviewer: ```suggestion
```

> ---

Reviewer 3: netlify is complaining about the file name: ```7:07:35 AM: WARN 2020/04/10 14:07:35 Content directory "/opt/build/repo/content/en/docs/setup/upgrade" have both index.* and _index.* files, pick one.
```

Author: That was present prior to this, although I can fix it in another pr

Reviewer 3: not needed thanks.

title: Upgrade Istio
description: Upgrade or downgrade Istio.
weight: 25
keywords: [kubernetes,upgrading]
---

## Canary upgrades

Upgrading Istio can be done by first running a canary deployment of the new control plane, allowing you
to monitor the effect of the upgrade with a small percentage of the workloads, before migrating all of the
traffic to the new version. This is much safer than doing an in place upgrade and is the recommended upgrade method.

When installing Istio, the `revision` installation setting can be used to deploy multiple independent control planes at the same time. A canary version of an upgrade can be started by installing the new Istio version's control plane next to the old one, using a different `revision` setting. Each revision is a full Istio control plane implementation with its own `Deployment`, `Service`, etc.

### Control plane

> If we want to install a new revision, which we call `canary`, we can must set the `revision` field:

Reviewer: nit: don't use "we" style writing in Istio docs (blogs excluded).
```suggestion
To install a new revision called `canary`, you would set the `revision` field as follows:
```
